### ISSUE TITLE
victorialogs: panics when using stats pipe

### ISSUE THREAD

**Issue by zekker6** [VictoriaMetrics/VictoriaMetrics#6306]
### Describe the bug

Panic occurs when using `stats` pipe.

### To Reproduce

1. Go to https://play-vmlogs.victoriametrics.com/select/vmui/

Run the following query: `_time: 10h * | unpack_json from _msg | stats count(level) asdf` a few times.
Observe the following panic:
```
goroutine 2039 [select]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.(*Table).pendingItemsFlusher(0xc00073b900)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset/table.go:733 +0xbc
github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.(*Table).startPendingItemsFlusher.func1()
    github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset/table.go:432 +0x1c
created by github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.(*Table).startPendingItemsFlusher in goroutine 1
    github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset/table.go:431 +0x65

goroutine 2040 [select]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.(*Table).inmemoryPartsFlusher(0xc00073b900)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset/table.go:748 +0xc5
github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.(*Table).startInmemoryPartsFlusher.func1()
    github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset/table.go:440 +0x1c
created by github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.(*Table).startInmemoryPartsFlusher in goroutine 1
    github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset/table.go:439 +0x65

goroutine 2041 [select]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.(*Table).startFlushCallbackWorker.func1()
    github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset/table.go:457 +0x7e
created by github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.(*Table).startFlushCallbackWorker in goroutine 1
    github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset/table.go:451 +0x6a

goroutine 2050 [select]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*datadb).inmemoryPartsFlusher(0xc0005eea80)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/datadb.go:278 +0xc5
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*datadb).startInmemoryPartsFlusher.func1()
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/datadb.go:268 +0x1c
created by github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*datadb).startInmemoryPartsFlusher in goroutine 1
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/datadb.go:267 +0x65

goroutine 2066 [select]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.(*Table).pendingItemsFlusher(0xc00073bb80)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset/table.go:733 +0xbc
github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.(*Table).startPendingItemsFlusher.func1()
    github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset/table.go:432 +0x1c
created by github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.(*Table).startPendingItemsFlusher in goroutine 1
    github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset/table.go:431 +0x65


goroutine 2067 [select]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.(*Table).inmemoryPartsFlusher(0xc00073bb80)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset/table.go:748 +0xc5
github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.(*Table).startInmemoryPartsFlusher.func1()
    github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset/table.go:440 +0x1c
created by github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.(*Table).startInmemoryPartsFlusher in goroutine 1
    github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset/table.go:439 +0x65

goroutine 1718 [IO wait]:
internal/poll.runtime_pollWait(0x7e92597a1680, 0x72)
    runtime/netpoll.go:345 +0x85
internal/poll.(*pollDesc).wait(0x7?, 0x1?, 0x0)
    internal/poll/fd_poll_runtime.go:84 +0x27
internal/poll.(*pollDesc).waitRead(...)
    internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Accept(0xc0002be000)
    internal/poll/fd_unix.go:611 +0x2ac
net.(*netFD).accept(0xc0002be000)
    net/fd_unix.go:172 +0x29
net.(*TCPListener).accept(0xc00016c0a0)
    net/tcpsock_posix.go:159 +0x1e
net.(*TCPListener).Accept(0xc00016c0a0)
    net/tcpsock.go:327 +0x30
github.com/VictoriaMetrics/VictoriaMetrics/lib/netutil.(*TCPListener).Accept(0xc0002be080)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/netutil/tcplistener.go:87 +0x3c
net/http.(*Server).Serve(0xc000622000, {0xb44c88, 0xc0002be080})
    net/http/server.go:3255 +0x33e
github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver.serveWithListener({0x9803eb, 0x5}, {0xb44c88, 0xc0002be080}, 0xac0928)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver/httpserver.go:170 +0x21e
github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver.serve({0x9803eb, 0x5}, 0x0, 0xac0928, 0x0)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver/httpserver.go:136 +0x5dd
created by github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver.Serve in goroutine 2079
    github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver/httpserver.go:106 +0x106

goroutine 2062 [select]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*datadb).inmemoryPartsFlusher(0xc0005eeb40)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/datadb.go:278 +0xc5
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*datadb).startInmemoryPartsFlusher.func1()
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/datadb.go:268 +0x1c
created by github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*datadb).startInmemoryPartsFlusher in goroutine 1
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/datadb.go:267 +0x65

goroutine 35514 [running]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*pipeStatsProcessorShard).getPipeStatsGroup(0xc000b06008, {0x0, 0x0, 0x7e9259682c28?})
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/pipe_stats.go:348 +0x208
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*pipeStatsProcessorShard).writeBlock(0xc000b06008, 0xc000b083c8)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/pipe_stats.go:206 +0x27b
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*pipeStatsProcessor).writeBlock(0xc000b083c8?, 0xc0004c94f0?, 0x2?)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/pipe_stats.go:395 +0x8c
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*pipeUnpackWriteContext).flush(0xc000b08380)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/pipe_unpack_logfmt.go:212 +0x6c
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*pipeUnpackJSONProcessor).writeBlock(0xc000626d50, 0xc0004fe018?, 0xc000148d88)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/pipe_unpack_json.go:108 +0x3c5
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*Storage).search.func1(0x2)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/storage_search.go:413 +0x222
created by github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*Storage).search in goroutine 35473
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/storage_search.go:398 +0x9b
```

2. Use the following query: `_time: 10h * | unpack_json from _msg | stats count(level) asdf` for logs which have json logs stored.
The following panic occurs:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x770e79]

goroutine 5054 [running]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.getBlockResultColumnIdxByName(...)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_result.go:1382
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*blockResult).getColumns(0xc0001398c8)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_result.go:1367 +0x259
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*blockResult).getColumnByName(0xc0001398c8, {0xc000db04b3, 0x5})
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_result.go:1336 +0x26
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*statsCountProcessor).updateStatsForAllRows(0xc004d91510, 0xc0001398c8)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/stats_count.go:50 +0xc8
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*pipeStatsGroup).updateStatsForAllRows(0xc001287688?, {0xc0002027d0, 0x1, 0x235?})
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/pipe_stats.go:361 +0x54
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*pipeStatsProcessorShard).writeBlock(0xc001287688, 0xc000139748)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/pipe_stats.go:207 +0x298
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*pipeStatsProcessor).writeBlock(0xc000139748?, 0xc0016b4848?, 0x2?)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/pipe_stats.go:395 +0x8c
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*pipeUnpackWriteContext).flush(0xc000139700)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/pipe_unpack_logfmt.go:212 +0x6c
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*pipeUnpackJSONProcessor).writeBlock(0xc0049fbd70, 0xc0050dd898?, 0xc00219a5a8)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/pipe_unpack_json.go:108 +0x3c5
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*Storage).search.func1(0x1)
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/storage_search.go:413 +0x222
created by github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*Storage).search in goroutine 3677
    github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/storage_search.go:398 +0x9b

```

### Version

v0.8.0

**Comment by valyala** [comment]
@zekker6 , thanks for the bug report! The bug should be fixed in the commit 22107421eb184765ae596c48d85bf357a2949e52 . This commit will be included in the next release. The root cause was in a data race when writing unpacked fields to the next pipe (stats in the case above). All the JSON unpack workers were using a single downstream shard for data writing, while every worker should read its own shard.

**Comment by valyala** [comment]
The issue has been fixed in VictoriaLogs [v0.9.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v0.9.0-victorialogs). Closing the issue then.
